src/runtime/graph_executor/graph_executor.cc

```diff
@@ -448,74 +570,65 @@ void GraphExecutor::SetupOpExecs() {
-      if (output_node_eids.count(input_eid) > 0) {
-        both_output_opinput_dltensors_[input_eid].push_back(
-            static_cast<DLTensor*>(op_args->arg_values[i].v_handle));
+    }
+    // check if any model output is the input of the op
+    if (output_node_eids.count(input_eid) > 0) {
+      auto& eid_both = both_output_opinput_dltensors_[input_eid];
+      auto both_tensor = static_cast<DLTensor*>(op_args->arg_values[i].v_handle);
+      if(std::find(eid_both.begin(), eid_both.end(), both_tensor) == eid_both.end()){
+        eid_both.push_back(both_tensor);
       }
     }
+  }
 
-    for (uint32_t i = inode.inputs.size(); i < inode.inputs.size() + inode.param.num_outputs; ++i) {
-      uint32_t output_eid = this->entry_id(nid, i - inode.inputs.size());
-      // check if op output is model output
-      if (output_node_eids.count(output_eid) > 0) {
-        output_dltensors_[output_eid].push_back(
-            static_cast<DLTensor*>(op_args->arg_values[i].v_handle));
+  for (uint32_t i = inode.inputs.size(); i < inode.inputs.size() + inode.param.num_outputs; ++i) {
+    uint32_t output_eid = this->entry_id(nid, i - inode.inputs.size());
+    // check if op output is model output
+    if (output_node_eids.count(output_eid) > 0) {
+      auto& eid_out = output_dltensors_[output_eid];
+      auto out_tensor = static_cast<DLTensor*>(op_args->arg_values[i].v_handle);
+      if(std::find(eid_out.begin(), eid_out.end(), out_tensor) == eid_out.end()){
+        eid_out.push_back(out_tensor);
       }
     }
   }
 }
-
 std::pair<std::function<void()>, std::shared_ptr<GraphExecutor::OpArgs> >
-GraphExecutor::CreateTVMOp(const TVMOpParam& param, const std::vector<DLTensor>& args) {
+GraphExecutor::CreateTVMOp(uint32_t nid, const TVMOpParam& param,
+                           const std::vector<DLTensor>& args) {
   std::shared_ptr<GraphExecutor::OpArgs> arg_ptr = std::make_shared<GraphExecutor::OpArgs>();
-  // setup address.
-  arg_ptr->args = args;
-  if (param.flatten_data) {
-    arg_ptr->shape_data.resize(arg_ptr->args.size());
-  }
-  for (size_t i = 0; i < arg_ptr->args.size(); ++i) {
-    TVMValue v;
-    DLTensor* t = &arg_ptr->args[i];
-    v.v_handle = t;
-    arg_ptr->arg_values.push_back(v);
-    arg_ptr->arg_tcodes.push_back(kTVMDLTensorHandle);
-    if (param.flatten_data) {
-      arg_ptr->shape_data[i] =
-          std::accumulate(t->shape, t->shape + t->ndim, 1, std::multiplies<int64_t>());
-      t->ndim = 1;
-      t->shape = &(arg_ptr->shape_data[i]);
-    }
-  }
+  // set up the memory location map
+  op_mem_location_map_[nid] = arg_ptr;
 
   if (param.func_name == "__nop") {
     return {[]() {}, arg_ptr};
   } else if (param.func_name == "__copy") {
     // Perform cross device data copy.
     // Directly copy data from the input to the output.
     // TODO(mbs): device_copy cleanup.
-    auto fexec = [arg_ptr]() {
+    auto fexec = [nid, this]() {
+      auto arg_ptr = this->op_mem_location_map_[nid];
```

Comment:
Perhaps it might be more economical to define `arg_ptr` outside the lambda and capture only that? (Same with the other lambdas that capture `this`.) I doubt it's a serious issue but it may be worth checking.